### PR TITLE
Update for running Android and iOS using AR Foundation.

### PR DESCRIPTION
**Pull request by TakahiroMiyaura**
## Overview
  I'm tring to develop Android and iPhone applications using MRTK V2.0.0,
  Using ARFoundation and MixedRealityInputSystem, I've been able to run the demo "Hand Interaction Exmples" on Android and iPhone.
  
 
## Changes
- #5875 .

### CODE REVIEW

**Comment by wiwei** [comment]
@davidkline-ms, ptal

**Comment by davidkline-ms** [comment]
> @davidkline-ms, ptal

On it!

**Comment by davidkline-ms** [comment]
@TakahiroMiyaura, thank you for the pull request! I will review and test it this week.

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs [hunk 1]
```diff
@@ -121,8 +136,8 @@ private void AddTouchController(Touch touch, Ray ray)
             }
 
             inputSystem?.RaiseSourceDetected(controller.InputSource, controller);
-            controller.StartTouch();
-            UpdateTouchData(touch, ray);
+            //controller.StartTouch();
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs]: Please remove

File: Assets/MobileDevices.SDK/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Movbile Devices SDK (Android / iPhone)
+This sdk is for mobile devices.Using ARFoundation,we can use 6dof head tracking!
+
+## Required ARFoundation Version
```
Comment (Troy-Ferrell) [inline Assets/MobileDevices.SDK/README.md]: Please move items to the appropriate existing MRTK.* folder. There is a documentation folder that would be great to have a how to setup MRTK project for mobile page :)

File: Assets/MobileDevices.SDK/Scenes/HandInteractionExamplesUsingARFoundation.unity.meta [hunk 3]
```diff
@@ -0,0 +1,7 @@
+fileFormatVersion: 2
+guid: d777fe74873e93144bde1bf89a98225c
```
Comment (Troy-Ferrell) [inline Assets/MobileDevices.SDK/Scenes/HandInteractionExamplesUsingARFoundation.unity.meta]: Same here. Please move to the  MRTK.Examples folder

File: ProjectSettings/ProjectVersion.txt [hunk 4]
```diff
@@ -1 +1,2 @@
-m_EditorVersion: 2018.4.6f1
+m_EditorVersion: 2019.1.13f1
```
Comment (davidkline-ms) [inline ProjectSettings/ProjectVersion.txt]: Please revert this change. This prevents the project from opening in Unity 2018.4. We have customers that require that version.

Thanks!

File: Assets/MobileDevices.SDK/Scripts/MobileDevicesCameraSystem.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+using System.Collections;
+using Microsoft.MixedReality.Toolkit;
+using Microsoft.MixedReality.Toolkit.CameraSystem;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using UnityEngine;
+using UnityEngine.SpatialTracking;
+using UnityEngine.XR.ARFoundation;
+
+namespace Com.Reseul.MobileDevices.CameraSystem
```
Comment (Troy-Ferrell) [inline Assets/MobileDevices.SDK/Scripts/MobileDevicesCameraSystem.cs]: Please use MRTK namespaces

File: ProjectSettings/ProjectSettings.asset [hunk 6]
```diff
@@ -3,7 +3,7 @@
 --- !u!129 &1
 PlayerSettings:
   m_ObjectHideFlags: 0
-  serializedVersion: 15
+  serializedVersion: 16
```
Comment (davidkline-ms) [inline ProjectSettings/ProjectSettings.asset]: Please revert this file as it is not related to the change.

Thanks!

File: ProjectSettings/GraphicsSettings.asset [hunk 7]
```diff
@@ -37,6 +37,7 @@ GraphicsSettings:
   - {fileID: 10770, guid: 0000000000000000f000000000000000, type: 0}
   - {fileID: 17000, guid: 0000000000000000f000000000000000, type: 0}
   - {fileID: 16000, guid: 0000000000000000f000000000000000, type: 0}
+  - {fileID: 16001, guid: 0000000000000000f000000000000000, type: 0}
```
Comment (davidkline-ms) [inline ProjectSettings/GraphicsSettings.asset]: Please revert this file as it is not related to the change.

Thanks!

File: ProjectSettings/EditorBuildSettings.asset [hunk 8]
```diff
@@ -6,6 +6,6 @@ EditorBuildSettings:
   serializedVersion: 2
   m_Scenes:
   - enabled: 1
-    path: Assets/MixedRealityToolkit.Examples/Demos/HandTracking/Scenes/HandInteractionExamples.unity
-    guid: 3dd4a396b5225f8469b9a1eb608bfa57
+    path: Assets/MobileDevices.SDK/Scenes/HandInteractionExamplesUsingARFoundation.unity
```
Comment (Troy-Ferrell) [inline ProjectSettings/EditorBuildSettings.asset]: Please remove all project asset files

File: Packages/manifest.json [hunk 9]
```diff
@@ -1,12 +1,18 @@
...
-    "com.unity.collab-proxy": "1.2.15",
-    "com.unity.package-manager-ui": "2.0.3",
-    "com.unity.purchasing": "2.0.3",
-    "com.unity.textmeshpro": "1.4.1",
-    "com.unity.xr.openvr.standalone": "1.0.2",
+    "com.unity.analytics": "3.3.2",
+    "com.unity.collab-proxy": "1.2.16",
+    "com.unity.multiplayer-hlapi": "1.0.2",
+    "com.unity.package-manager-ui": "2.1.2",
+    "com.unity.purchasing": "2.0.6",
```
Comment (Troy-Ferrell) [inline Packages/manifest.json]: ARFoundation related items may need to be added but the others should be removed

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchController.cs [hunk 10]
```diff
@@ -14,6 +14,23 @@ public class UnityTouchController : BaseController
         public UnityTouchController(TrackingState trackingState, Handedness controllerHandedness, IMixedRealityInputSource inputSource = null, MixedRealityInteractionMapping[] interactions = null)
                 : base(trackingState, controllerHandedness, inputSource, interactions)
         {
+            if (InputSystem?.InputSystemProfile.GesturesProfile != null)
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchController.cs]: Why did this move out of `SetupDefaultInteractions`? Is that not being called or not being called reliably?

File: Assets/MobileDevices.SDK/Scripts/MobileDevicesCameraSystem.cs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+{
+    public class MobileDevicesCameraSystem : BaseCoreSystem, IMixedRealityCameraSystem
+    {
+        private ARCameraBackground _aRCameraBackground;
```
Comment (Troy-Ferrell) [inline Assets/MobileDevices.SDK/Scripts/MobileDevicesCameraSystem.cs]: please remove _ for private variables. Not consistent with rest of codebase

File: Assets/MobileDevices.SDK/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
+# Movbile Devices SDK (Android / iPhone)
```
Comment (keveleigh) [inline Assets/MobileDevices.SDK/README.md]:
```suggestion
# Mobile Devices SDK (Android / iPhone)
```

File: Assets/MobileDevices.SDK/Scripts/MobileDevicesCameraSystem.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        private void ReleaseForARFoundation()
+        {
+            if (MixedRealityPlayspace.Transform.GetComponentInChildren<ARSessionOrigin>() == null) return;
+            if (MixedRealityPlayspace.Transform.GetComponentInChildren<ARSession>() == null) return;
+
+            CameraCache.Main.transform.parent = MixedRealityPlayspace.Transform;
+            Object.DestroyImmediate(_trackedPoseDriver);
+            Object.DestroyImmediate(_aRCameraBackground);
```
Comment (Troy-Ferrell) [inline Assets/MobileDevices.SDK/Scripts/MobileDevicesCameraSystem.cs]: I believe this should be Object.Destroy?

File: Assets/MobileDevices.SDK/README.md (hunk 12 above)
Comment (davidkline-ms) [inline Assets/MobileDevices.SDK/README.md]: We should consider in which folder these components belong. Quite possibly many should be folded into the MixedRealityToolkit.* folders.

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs [hunk 14]
```diff
@@ -50,8 +50,10 @@ public override void Update()
                     case TouchPhase.Began:
                         AddTouchController(touch, ray);
                         break;
-                    case TouchPhase.Moved:
                     case TouchPhase.Stationary:
+                        StartTouchData(touch,ray);
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs]: I'm not sure I understand this change. Why would we start the touch when the touch phase is stationary (which I think can happen multiple times during a touch) instead of the existing start when a touch controller is added?

https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5876/files#diff-11e52ac5aede93183e1852123ffc04cbL124

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs (hunk 1 above)
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs]: nit: please remove the commented lines

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchController.cs (hunk 10 above)
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchController.cs]: Question: Does this code not work properly if left in SetupDefaultInteractions()?

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs [hunk 15]
```diff
@@ -67,6 +69,19 @@ public override void Update()
...
+        private void StartTouchData(Touch touch,Ray ray)
+        {
+            UnityTouchController controller;
+
+            if (!ActiveTouches.TryGetValue(touch.fingerId, out controller))
+            {
+                return;
+            }
+
+            controller.StartTouch();
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs]: this will cause a pointerdown and a holdstarted for every stationary touch.

**Comment by TakahiroMiyaura** [comment]
Thank you for feedback and reviewing!
I've been confirming the parts you pointed out.
Please give me a little more time.

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchController.cs (hunk 10 above)
Comment (Pycorax) [reply Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchController.cs]: Not the original submitter, but I've been testing this and it seems like it works fine when left in SetupDefaultInteractions().

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs (hunk 14 above)
Comment (Pycorax) [reply Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs]: I've been experimenting with this part of the code and it seems like dragging objects in the sample scene only works if the InputSystem's RaisePOinterDown event is called during TouchPhase.Stationary case in this code block.

I'm not familiar enough with MRTK to understand why but I think this is why @TakahiroMiyaura added that function call there.

**Comment by Pycorax** [comment]
> 
> 
> Thank you for feedback and reviewing!
> I've been confirming the parts you pointed out.
> Please give me a little more time.

I have looked into the code as well since I require this feature too. I have some of these fixes over at https://github.com/Pycorax/MixedRealityToolkit-Unity/commit/2541a607d69e4cc7bdc10beed961dc170e7e53b5

I hope this helps :D

File: Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs (hunk 15 above)
Comment (TakahiroMiyaura) [reply Assets/MixedRealityToolkit/Providers/UnityInput/UnityTouchDeviceManager.cs]: Sorry...
You're completely right.
This should call first time only.

**Comment by emartinDesignInteractive** [comment]
I was able to integrate this into my project, however I notice that a blue ray is being drawn anytime i am interacting with interactables, anyone else seen this? Seems to only happen on the device and not the editor.

Thanks!

**Comment by TakahiroMiyaura** [comment]
Hello, 
I'm sorry to have kept you waiting. 
I've been confirming and researching yet... 
I'm in trouble.
Now I've been understanding  the below. 
  
* Invoking "controller.StartTouch()" within UnityTouchDeviceManager.AddTouchController, TouchPointer.Result property is not yet set.So we can't get the focus objects. 
* Invoking "controller.StartTouch()" after executing  UnityTouchDeviceManager.AddTouchController, TouchPointer.Result property is set the value, and We can get the focus objects.So we can use pointerdown event. 

Updating to invoke "controller.StartTouch()" only once on stationary touch, we will be able to resolve this problem. 
But I doubt that this update isn't a solution of the root cause.
 
Would you please advise me which I should choosing the below?
  
1. I update to invoke "controller.StartTouch()" only once on stationary touch. 
1. Researching this problem,and I find out the root cause(I think thar it may be more time). 

Thanks.

**Comment by davidkline-ms** [comment]
Thanks for the submission @TakahiroMiyaura! Your work was adapted into PR #6344.
